crates/rpc/src/server/accept.rs

```diff
@@ -150,16 +179,35 @@ impl AcceptHeaderLayer {
                 continue;
             }
 
-            // Skip if the genesis commitment does not match.
+            // Skip if the genesis commitment does not match, or if it is required but missing.
             let genesis = media_type
                 .get_param(Self::GENESIS)
                 .map(|value| Word::try_from(value.unquoted_str().as_ref()))
                 .transpose()
                 .map_err(AcceptHeaderError::InvalidGenesis)?;
-            if let Some(genesis) = genesis
-                && genesis != self.genesis_commitment
-            {
-                continue;
+
+            match genesis_mode {
+                GenesisNegotiation::Mandatory => {
+                    match genesis {
+                        None => {
+                            // For write methods, the genesis parameter must be present.
+                            continue;
+                        },
+                        Some(genesis) if genesis != self.genesis_commitment => {
+                            // Present but mismatched.
+                            continue;
+                        },
+                        _ => {},
+                    }
+                },
+                GenesisNegotiation::Optional => {
+                    if let Some(genesis) = genesis
+                        && genesis != self.genesis_commitment
+                    {
+                        // Present but mismatched.
+                        continue;
+                    }
+                },
```

Comment:
Nit: we can simplify by collapsing conditions a bit


```suggestion
            match (genesis_mode, genesis) {
                (_, Some(value)) if value != self.genesis_commitment => continue,
                (GenesisNegotiation::Mandatory, None) => continue,
                _ => {},
            }
```